caddyhttp/proxy/proxy.go

```diff
@@ -231,6 +224,16 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 		if !keepRetrying() {
 			break
 		}
+
+		// Before we retry the request we have to make sure
+		// that the body is rewound to it's beginning.
+		//
+		// NOTE:
+		//   Reaching this point implies that keepRetrying() above returned true,
+		//   which in turn implies that body must be a *bufferedBody.
+		if err := body.(*bufferedBody).rewind(); err != nil {
```

Comment:
@mholt Fixed.

Can I squash the last 3 commits into a single one before you merge it to clean up the history? It'd result in a force push though of course.